README

RHid  -- Ruby USB HID class access library

NISHI Takao <zophos@koka-in.org>

= What's RHid

RHid is Ruby USB HID class access library.

= Requirements

* Windows XP/vista/2003
* Ruby 1.8 or higher (MSWIN/Cygwin)

= Copying

Copyright (c) 2009 NISHI Takao <zophos@koka-in.org> All rights reserved.
This is free software with ABSOLUTELY NO WARRANTY.

You can redistribute it and/or modify it under the terms of Ruby's licence.

Comments, suggestions, patches and beers are welcome.


ChangeLog
2010-03-16  NISHI Takao  <zophos@koka-in.org>

	* src/rhid/win32.rb (RHid::Win32::Api): SetupDiGetClassDevsA() 2nd
	argument changed from char * to void * to avoid error on Windows7. 
	(RHid::Win32#open_by_id, etc): substitute DL.malloc for DL.strdup.

2009-09-21  NISHI Takao  <zophos@koka-in.org>

	* src/usbio.rb: made thread safe

	* src/mr999.rb: modified some action timings.
	  - MR999::grip: renames to MR999::grasp. 

2009-09-21  NISHI Takao  <zophos@koka-in.org>
	
	* src/rhid/win32.rb (RHid::Win32::Api): fixed ReadFile() 2nd
	argument type.

	* src/usbio.rb (UsbIo::Port::from_i, UsbIo::read): implemented

	* src/rhid.rb (RHid::<<): changed from alias to method.
	RHid#<< returns self.

2009-09-20  NISHI Takao  <zophos@koka-in.org>

	* src/rhid/win32.rb (RHid::Win32::open_by_id): fixed typo

2009-09-20  NISHI Takao  <zophos@koka-in.org>

	* src/rhid/win32.rb (RHid::Win32::open_by_id): fixed
	Api.setupDiEnumDeviceInterfaces() bug

	* src/rhid.rb: added some comments

	* ChangeLog: added


src/rhid.rb
#
# rhid.rb -- HID class access library
#
# NISHI Takao <zophos@koka-in.org>
#
class RHid
    class DoRetryError<StandardError;end

    Capabilities=Struct.new(:usage,
                            :usagePage,
                            :inputReportByteLength,
                            :outputReportByteLength,
                            :featureReportByteLength,
                            :numberLinkCollectionNodes,
                            :numberInputButtonCaps,
                            :numberInputValueCaps,
                            :numberInputDataIndices,
                            :numberOutputButtonCaps,
                            :numberOutputValueCaps,
                            :numberOutputDataIndices,
                            :numberFeatureButtonCaps,
                            :numberFeatureValueCaps,
                            :numberFeatureDataIndices)

    def self.open_by_id(vendor_id,product_id,&block)
        self.new.open_by_id(vendor_id,product_id,&block)
    end

    def self.open_by_path(path,&block)
        self.new.open_by_id(path,&block)
    end

    def initialize
        @capabilities=nil
        super
    end
    def <<(data)
        self.write
        self
    end
end


src/mr999.rb
#
# mr999.rb --  ELEKIT MR-999 and IF-100 accesse library
#
# NISHI Takao <zophos@koka-in.org>
#
require 'usbio'


class MR999
    WAIST=0
    SHOULDER=1
    ELBOW=2
    WRITE=3
    FINGER=4

    ######################################################################
    #
    #
    #
    class Joint
        def initialize(usbio,port,nornaml_bit,reverse_bit)
            @usbio=usbio
            @port=port
            @normal_bit=nornaml_bit
            @reverse_bit=reverse_bit
        end
        def normal(exec=false)
            @usbio.port[@port][@normal_bit]=UsbIo::ON
            @usbio.port[@port][@reverse_bit]=UsbIo::OFF
            self.exec if exec
        end
        def reverse(exec=false)
            @usbio.port[@port][@normal_bit]=UsbIo::OFF
            @usbio.port[@port][@reverse_bit]=UsbIo::ON
            self.exec if exec
        end
        def toggle(exec=false)
            @usbio.port[@port].toggle(@normal_bit)
            @usbio.port[@port].toggle(@reverse_bit)
            self.exec if exec
        end
        def stop(exec=false)
            @usbio.port[@port][@normal_bit]=UsbIo::OFF
            @usbio.port[@port][@reverse_bit]=UsbIo::OFF
            self.exec if exec
        end
        def exec
            @usbio.write(@port)
        end
    end
    #
    # end of MR999::Joint
    #

    def initialize(vendor_id=0x12ed,product_id=0x1003)
        @usbio=UsbIo.new(vendor_id,product_id)

        @joint=[@waist=Joint.new(@usbio,0,5,4),
            @shoulder=Joint.new(@usbio,0,6,7),
            @elbow=Joint.new(@usbio,0,3,2),
            @wrist=Joint.new(@usbio,1,1,0),
            @finger=Joint.new(@usbio,0,0,1)].freeze
    end
    attr_reader :waist,:shoulder,:elbow,:wrist,:finger

    def exec
        @usbio.write
    end

    def stop
        @joint.each{|j| j.stop(false) }
        @usbio.write
    end

    #
    # set initial position
    #
    def init
        self.stop

        $stderr<<"shoulder"
        self.shoulder.normal(true)
        STDIN.getc
        self.stop
        
        $stderr<<"elbow"
        self.elbow.reverse(true)
        STDIN.getc
        self.stop
        
        $stderr<<"wrist"
        self.wrist.normal(true)
        STDIN.getc
        self.stop
        
        $stderr<<"finger"
        self.finger.normal(true)
        STDIN.getc
        self.stop

        $stderr<<"waist"
        self.waist.normal(true)
        STDIN.getc
        self.stop

        $stderr<<"\n"
        
        self
    end

    #
    # move from initial position to grasping position, and grasp something
    #
    # You MUST adjust each sleep length with your MR-999
    #
    def grasp
        self.stop

        #
        # rotate the wasit and the wrist
        #
        self.waist.reverse(true)
        sleep(14.5)
        self.wrist.reverse(true)
        sleep(5.5)
        self.stop

        #
        # extend the elbow and shoulder, and open the finger
        #
        self.elbow.normal(true)
        sleep(15)
        self.shoulder.reverse(true)
        sleep(12)
        self.finger.reverse(true)
        sleep(3)
        self.stop

        #
        # waki-waki
        #
        2.times{|i|
            self.finger.normal(true)
            sleep(0.5)
            self.finger.reverse(true)
            sleep(0.5)
        }
        self.stop


        $stderr<<"Hit Enter to Continue..."
        STDIN.getc

        
        #
        # close finger
        #
        self.finger.normal(true)
        sleep(3)
        self.stop

        #
        # shrink the elbow and the shoulder 
        #
        self.shoulder.normal
        self.elbow.reverse
        self.exec
        sleep(5)
        self.stop

        self
    end

    #
    # shake the grasping somthing (e.g, leek)
    #
    def shake
        self.stop

        self.waist.normal(true)
        self.wrist.normal(true)
        sleep(1.5)

        5.times{|i|
            self.waist.toggle(true)
            self.wrist.toggle(true)
            sleep(3)
        }
        self.stop

        self
    end
end


src/rhid/win32.rb
#
# rhid/win32.rb -- HID class access library for Win32API
#
# NISHI Takao <zophos@koka-in.org>
#
require 'rhid'

require 'dl/import'
require 'dl/struct'

class RHid
    module Win32

        ##################################################################
        #
        # API Declaring
        #
        module Api
            extend DL::Importable
            
            GENERIC_READ=[0x80000000].pack('L').unpack('l')[0]
            GENERIC_WRITE=[0x40000000].pack('L').unpack('l')[0]
            FILE_SHARE_READ=0x01
            FILE_SHARE_WRITE=0x02
            
            OPEN_EXISTING=3
            
            DIGCF_PRESENT=0x02
            DIGCF_DEVICEINTERFACE=0x10
            
            HidP_Input=0
            HidP_Output=1
            HidP_Feature=2
            
            
            dlload('kernel32.dll',
                   'hid.dll',
                   'setupapi.dll')
            
            GUID=
                struct([
                           'long  data1',
                           'short data2',
                           'short data3',
                           'char  data4[8]'
                       ])
            
            HIDD_ATTRIBUTES=
                struct([
                           'long cbSize',
                           'short vendorID',
                           'short productID',
                           'short versionNumber'
                       ])
            
            HIDP_CAPS=
                struct([
                           'short usage',
                           'short usagePage',
                           'short inputReportByteLength',
                           'short outputReportByteLength',
                           'short featureReportByteLength',
                           'short reserved[17]',
                           'short numberLinkCollectionNodes',
                           'short numberInputButtonCaps',
                           'short numberInputValueCaps',
                           'short numberInputDataIndices',
                           'short numberOutputButtonCaps',
                           'short numberOutputValueCaps',
                           'short numberOutputDataIndices',
                           'short numberFeatureButtonCaps',
                           'short numberFeatureValueCaps',
                           'short numberFeatureDataIndices'
                       ])
            
            HidP_Value_Caps=
                struct([
                           'short usagePage',
                           'char * reportID',
                           'long isAlias',
                           'short bitField',
                           'short linkCollection',
                           'short linkUsage',
                           'short linkUsagePage',
                           'long isRange',
                           'long isStringRange',
                           'long isDesignatorRange',
                           'long isAbsolute',
                           'long hasNull',
                           'char * reserved',
                           'short bitSize',
                           'short reportCount',
                           'short reserved2',
                           'short reserved3',
                           'short reserved4',
                           'short reserved5',
                           'short reserved6',
                           'long logicalMin',
                           'long logicalMax',
                           'long physicalMin',
                           'long physicalMax',
                           'short usageMin',
                           'short usageMax',
                           'short stringMin',
                           'short stringMax',
                           'short designatorMin',
                           'short designatorMax',
                           'short dataIndexMin',
                           'short dataIndexMax'
                       ])
            
            SP_DEVICE_INTERFACE_DATA=
                struct([
                           'long cbSize',
                           'long  interfaceClassGuid_Data1',
                           'short interfaceClassGuid_Data2',
                           'short interfaceClassGuid_Data3',
                           'char  interfaceClassGuid_Data4[8]',
                           'long flags',
                           'long reserved'
                       ])
            
            SP_DEVICE_INTERFACE_DETAIL_DATA=
                struct([
                           'long cbSize',
                           'char devicePath[1]'
                       ])
            
            SP_DEVINFO_DATA=
                struct([
                           'long cbSize',
                           'long  classGuid_Data1',
                           'short classGuid_Data2',
                           'short classGuid_Data3',
                           'char  classGuid_Data4[8]',
                           'long devInst',
                           'long reserved'
                       ])
            

            #
            # kernel32.dll
            #
            extern(<<_EOS_
long CreateFileA(char *,long,long,long *,long,long,long)
_EOS_
                   )

            extern(<<_EOS_
long CloseHandle(long)
_EOS_
                   )

            extern(<<_EOS_
long ReadFile(long,void *,long,long *,long)
_EOS_
                   )

            extern(<<_EOS_
long WriteFile(long,char *,long,long *,long)
_EOS_
                   )

            extern(<<_EOS_
long FormatMessageA(long,void *,long,long,char *,long,long)
_EOS_
                   )
            extern(<<_EOS_
long GetLastError()
_EOS_
                   )


            #
            # hid.dll
            #
            extern(<<_EOS_
long HidD_FreePreparsedData(long)
_EOS_
                   )

            extern(<<_EOS_
long HidD_GetAttributes(long,HIDD_ATTRIBUTES *)
_EOS_
                   )

            extern(<<_EOS_
long HidD_GetHidGuid(GUID *)
_EOS_
                   )

            extern(<<_EOS_
long HidD_GetPreparsedData(long,long *)
_EOS_
                   )

            extern(<<_EOS_
long HidP_GetCaps(long,HIDP_CAPS *)
_EOS_
                   )

            extern(<<_EOS_
long HidP_GetValueCaps(short,char *,short *,long)
_EOS_
                   )

            #
            # setupapi.dll
            #
            extern(<<_EOS_
long SetupDiCreateDeviceInfoList(GUID *,long)
_EOS_
                   )

            extern(<<_EOS_
long SetupDiDestroyDeviceInfoList(long)
_EOS_
                   )

            extern(<<_EOS_
long SetupDiEnumDeviceInterfaces(long,
    long,
    GUID *,
    long,
    SP_DEVICE_INTERFACE_DATA *)
_EOS_
                   )

            extern(<<_EOS_
long SetupDiGetClassDevsA(GUID *,void *,long,long)
_EOS_
                   )

            extern(<<_EOS_
long SetupDiGetDeviceInterfaceDetailA(long,
    SP_DEVICE_INTERFACE_DATA *,
    SP_DEVICE_INTERFACE_DETAIL_DATA *,
    long,
    long *,
    long)
_EOS_
                   )
        end
        #
        # end of RHid::Win32::Api
        #
    

        ##################################################################
        #
        #
        #
        def initialize
            @path=nil
            @handle=nil
        end

        ##################################################################
        #
        #
        #
        def open_by_id(vendor_id,product_id,&block)
            raise(IOError,'Already opened') if @handle

            guid=Api::GUID.malloc
            Api.hidD_GetHidGuid(guid)
            dis=Api.setupDiGetClassDevsA(guid,nil,0,
                                         Api::DIGCF_PRESENT|
                                             Api::DIGCF_DEVICEINTERFACE)

            did=Api::SP_DEVICE_INTERFACE_DATA.malloc
            did.cbSize=Api::SP_DEVICE_INTERFACE_DATA.size
            
            sz=DL.malloc(DL::sizeof('L'))
            
            hidda=Api::HIDD_ATTRIBUTES.malloc
            hidda.cbSize=hidda.size
            
            i=0
            begin
                raise Errno::ENOENT if Api.setupDiEnumDeviceInterfaces(dis,
                                                                       0,
                                                                       guid,
                                                                       i,
                                                                       did
                                                                       )==0
                ret=
                    Api.setupDiGetDeviceInterfaceDetailA(dis,did,nil,0,sz,0)
                s=sz.to_a('L')[0]
                didd=Api::SP_DEVICE_INTERFACE_DETAIL_DATA.malloc(s)
                didd.cbSize=5
                
                ret=
                    Api.setupDiGetDeviceInterfaceDetailA(dis,did,didd,s,nil,0)
                raise DoRetryError if ret==0
                
                @path=didd.to_ptr[4,s-didd.cbSize]
                @handle=Api.createFileA(@path,
                                        Api::GENERIC_READ|
                                            Api::GENERIC_WRITE,
                                            Api::FILE_SHARE_READ|
                                            Api::FILE_SHARE_WRITE,
                                        nil,
                                        Api::OPEN_EXISTING,
                                        0,0)
                raise DoRetryError if @handle==-1
                
                ret=Api.hidD_GetAttributes(@handle,hidda)
                
                raise DoRetryError if (ret==0 ||
                                           hidda.vendorID!=vendor_id ||
                                           hidda.productID!=product_id)
            rescue DoRetryError
                Api.closeHandle(@handle) if (@handle&&@handle!=-1)
                @handle=nil
                @path=nil
                i+=1
                retry
            rescue Errno::ENOENT
                @handle=nil
                @path=nil
                raise
            ensure
                Api.setupDiDestroyDeviceInfoList(dis)
            end
            
            ObjectSpace.define_finalizer(self,RHid.finalize(@handle))
            if(block)
                yield(self)
                self.close
            else
                self
            end
        end
        #
        # end of open_by_id
        #


        ##################################################################
        #
        #
        #
        def open_by_path(path,&block)
            raise(IOError,'Already opened') if @handle

            @handle=Api.createFileA(path,
                                    Api::GENERIC_READ|
                                        Api::GENERIC_WRITE,
                                    Api::FILE_SHARE_READ|
                                        Api::FILE_SHARE_WRITE,
                                    nil,
                                    Api::OPEN_EXISTING,
                                    0,0)
            if(@handle==-1)
                @hadnle=nil
                raise Errno::ENOENT
            end
            
            @path=path
            ObjectSpace.define_finalizer(self,RHid.finalize(@handle))
            if(block)
                yield(self)
                self.close
            else
                self
            end
        end
        #
        # end of open_by_path
        #

        ##################################################################
        #
        #
        #
        def close
            ObjectSpace.undefine_finalizer(self)

            Api.closeHandle(@handle) if @handle
            @capabilities=nil
            @handle=nil
        end
        #
        # end of close
        #


        ##################################################################
        #
        #
        #
        def capabilities
            raise IOError unless @handle
            return @capabilities if @capabilities

            preparsed_data=DL.malloc(DL::sizeof('L'))
            if(Api.hidD_GetPreparsedData(@handle,preparsed_data)==1)
                caps=Api::HIDP_CAPS.malloc
                unless(Api.hidP_GetCaps(preparsed_data.to_a('L')[0],caps)==0)
                    @capabilities=Capabilities.new(caps.usage,
                                                   caps.usagePage,
                                                   caps.inputReportByteLength,
                                                   caps.outputReportByteLength,
                                                   caps.featureReportByteLength,
                                                   caps.numberLinkCollectionNodes,
                                                   caps.numberInputButtonCaps,
                                                   caps.numberInputValueCaps,
                                                   caps.numberInputDataIndices,
                                                   caps.numberOutputButtonCaps,
                                                   caps.numberOutputValueCaps,
                                                   caps.numberOutputDataIndices,
                                                   caps.numberFeatureButtonCaps,
                                                   caps.numberFeatureValueCaps,
                                                   caps.numberFeatureDataIndices)

                end

                #
                # fixme: implement Api.hidP_GetValueCaps 
                #

                Api::hidD_FreePreparsedData(preparsed_data.to_a('L')[0])
            end
            

            @capabilities
        end
        #
        # end of capabilities
        #


        ##################################################################
        #
        #
        #
        def read
            self.capabilities
            raise IOError unless @capabilities

            buf=DL.malloc(@capabilities.inputReportByteLength)
            sz=DL.malloc(DL::sizeof('L'))
            ret=Api.readFile(@handle,
                             buf,
                             @capabilities.inputReportByteLength,
                             sz,0)

            raise(IOError,"#{Api.getLastError}") if ret==0

            buf.to_s(sz.to_a('L')[0])
        end
        #
        # end of read
        #

        ##################################################################
        #
        #
        #
        def write(data)
            self.capabilities
            raise IOError unless @capabilities

            data=data.to_s
            sz=data.unpack('C*').size
            if(sz<@capabilities.outputReportByteLength)
                data+="\xff"*(@capabilities.outputReportByteLength-sz)
            end

            sz=DL.malloc(DL::sizeof('L'))
            ret=Api.writeFile(@handle,
                              data,
                              @capabilities.outputReportByteLength,
                              sz,0)
            raise(IOError,"#{Api.getLastError}") if ret==0

            sz.to_a('L')[0]
        end
        #
        # end of write
        #
    end
    #
    # end of RHid::Win32
    #

    include Win32

    def RHid.finalize(handle)
        Proc.new{
            Api.closeHandle(handle)  if (handle && handle!=-1)
        }
    end

end


src/usbio.rb
#
# usbio.rb -- wrapper for Morphy Planning USB-IO family
#
# NISHI Takao <zophos@koka-in.org>
#
require 'rhid/win32'
require 'thread'

class UsbIo
    ON=Lo=0
    OFF=Hi=1
    
    class Port
        def initialize(bit_num=8,init_val=Hi)
            @bits=Array.new(bit_num)
            self.fill(init_val)
        end
        def fill(n)
            raise ArgumentError if(n!=Lo&&n!=Hi)
            @bits.map!{|x| x=n }
        end
        def [](x)
            @bits[x]
        end
        def []=(x,n)
            x=x.to_i
            raise ArgumentError if((n!=Lo&&n!=Hi)||(x>=@bits.size))
            @bits[x]=n
        end
        def toggle(x)
            raise ArgumentError if(x>=@bits.size)
            @bits[x]^=Hi
        end

        def to_s
            [@bits.join('')].pack('b*')
        end
        def to_i
            self.to_s.unpack('C*')[0]
        end
        def to_a
            @bits.dup
        end

        def from_i(val)
            @bits=[val.to_i].pack('C*').unpack('b*')[0].split(//).map{|x|
                x.to_i
            }[0...@bits.size]
        end
    end

    def initialize(vendor_id=0x12ed,product_id=0x1003)
        @port=[Port.new(8),Port.new(4)].freeze
        @mutex=Mutex.new
        @seq=1
        self.open(vendor_id,product_id)
    end
    attr_reader :port

    def open(vendor_id=0x12ed,product_id=0x1003)
        @connector=RHid.open_by_id(vendor_id,product_id) unless @connector
        @port.each{|p| p.fill(OFF) }
        self.write
        self
    end
    def close
        @connector.close if @connector
        @connector=nil
    end

    def write(port=nil)
        cmd=nil
        case port
        when nil
            return self.write(0)+self.write(1)
        when 0
            cmd=0x01
        when 1
            cmd=0x02
        else
            raise ArgumentError
        end

        ret=nil
        @mutex.synchronize{
            ret=@connector.write(_build_request(cmd,@port[port]))
        }

        ret
    end

    def read(port=nil)
        cmd=nil
        case port
        when nil
            return self.read(0)+self.read(1)
        when 0
            cmd=0x03
        when 1
            cmd=0x04
        else
            raise ArgumentError
        end

        val=nil
        @mutex.synchronize{
            @connector.write(_build_request(cmd,0xff))

            #
            # compare command and sequence number between the query with
            # the reply to discard pre-fetched replies.
            #
            begin
                (rcmd,val,oid,seq)=_parse_reply(@connector.read)
                raise RHid::DoRetryError if((rcmd!=cmd)||
                                                (oid!=self.__id__)||
                                                (seq!=@seq))
            rescue RHid::DoRetryError
                retry
            end
        }

        @port[port].from_i(val)
    end

    private
    def _build_request(cmd,val)
        #
        # set sequence number to the request message
        #
        @seq=[@seq+1].pack('n').unpack('n')[0]
        [cmd.to_i,val.to_i,self.__id__,@seq].pack('nCNn')
    end

    def _parse_reply(rpl)
        rpl.unpack('nCNn')
    end
end
